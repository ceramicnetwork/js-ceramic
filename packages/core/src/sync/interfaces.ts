import type { CID } from 'multiformats/cid'
import type { StreamID } from '@ceramicnetwork/streamid'
import type { SupportedNetwork } from '@ceramicnetwork/anchor-utils'

export type SyncConfig = {
  /**
   * Database connection string.
   */
  db: string
  chainId: SupportedNetwork
}

export interface ISyncApi {
  startModelSync(models: string | string[], startBlock?: number, endBlock?: number): Promise<void>
  shutdown(): Promise<void>
}

// handles a commit found during a sync
// should be similar to: https://github.com/ceramicnetwork/js-ceramic/blob/6ae6e121b33132225f256762825e1439fd84f23a/packages/core/src/state-management/state-manager.ts#L210
export type HandleCommit = (streamId: StreamID, commit: CID, model?: StreamID) => Promise<void>

export interface IpfsService {
  retrieveFromIPFS(cid: CID | string, path?: string): Promise<any>
  retrieveCommit(cid: CID | string, streamId: StreamID): Promise<any>
  storeCommit(data: any, streamId?: StreamID): Promise<CID>
  storeRecord(record: Record<string, unknown>): Promise<CID>
}

export interface TreeMetadata {
  numEntries: number
  streamIds: string[]
}

export const REBUILD_ANCHOR_JOB = 'rebuildAnchorJob'
export interface RebuildAnchorJobData {
  models: string[]
  chainId: string
  txHash: string
  root: string
  txType?: string
}

export const HISTORY_SYNC_JOB = 'historySyncJob'
export const CONTINUOUS_SYNC_JOB = 'continuousSyncJob'

export type SyncJob = typeof HISTORY_SYNC_JOB | typeof CONTINUOUS_SYNC_JOB
export interface SyncJobData {
  fromBlock: number
  toBlock: number
  models: string[]
}

export type JobData = RebuildAnchorJobData | SyncJobData
