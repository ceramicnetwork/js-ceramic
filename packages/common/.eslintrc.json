{
  "root": true,
  "parser": "@typescript-eslint/parser",
  "plugins": ["@typescript-eslint", "import"],
  "rules": {
    "@typescript-eslint/no-explicit-any": "off",
    "@typescript-eslint/ban-ts-ignore": "off",
    "@typescript-eslint/ban-ts-comment": "off",
    "import/no-unresolved": "error",
    "import/default": "off"
  },
  "settings": {
    "import/parsers": {
      "@typescript-eslint/parser": [".ts", ".tsx"]
    },
    "import/resolver": {
      "typescript": {
        "extensions": []
      },
      "node": true
    }
  },
  "extends": [
    "eslint:recommended",
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:import/typescript"
  ]
}
