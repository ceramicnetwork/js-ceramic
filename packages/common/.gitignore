node_modules/
coverage/
lib/
src/generated/
