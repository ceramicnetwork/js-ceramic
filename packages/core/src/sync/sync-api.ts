import type { CID } from 'multiformats/cid'
import { StreamID } from '@ceramicnetwork/streamid'

export interface ISyncApi {}

interface IpfsService {
  retrieveFromIPFS(cid: CID | string, path?: string): Promise<any>
  storeCommit(data: any, streamId?: StreamID): Promise<CID>
  storeRecord(record: Record<string, unknown>): Promise<CID>
}

interface Protocol {
  // will rename but this is the function we want to use
  handlePubsubUpdate(streamId: StreamID, tip: CID, model?: StreamID): Promise<void>
}

export class SyncApi implements ISyncApi {
  constructor(private readonly ipfsService: IpfsService, private readonly protocol: Protocol) {
    // create job queue
  }

  async init(): Promise<void> {
    // initialize job queue with handlers from ./workers
  }
}
