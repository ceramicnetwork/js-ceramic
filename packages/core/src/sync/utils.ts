import { CID } from 'multiformats/cid'
import { IpfsService } from './interfaces.js'

const METADATA_PATH = '2'

interface Metadata {
  numEntries: number
  streamIds: string[]
}

enum PathDirection {
  L,
  R,
}

const getPath = (index: number, leavesCount: number): PathDirection[] => {
  if (index >= leavesCount) {
    throw Error(`Leaf at index ${index} does not exist as there are only ${leavesCount} leaves`)
  }

  // only one leaf in the entire tree
  if (leavesCount === 1) {
    return [PathDirection.L]
  }

  return getPathHelper(index, leavesCount)
}

const getPathHelper = (index: number, leavesCount: number, path = []): PathDirection[] => {
  if (leavesCount <= 1) {
    return path
  }

  if (leavesCount === 2) {
    index === 0 ? path.push(PathDirection.L) : path.push(PathDirection.R)
    return path
  }

  const middleIndex = Math.trunc(leavesCount / 2)
  if (index < middleIndex) {
    path.push(PathDirection.L)
    return getPathHelper(index, middleIndex, path)
  }

  path.push(PathDirection.R)
  return getPathHelper(index - middleIndex, leavesCount - middleIndex, path)
}

/**
 * Retrieves the leaves of a Merkle Tree based on the merkle trees metadata
 */
export class MerkleTreeLoader {
  private _metadata: Metadata
  constructor(private readonly ipfsService: IpfsService, private readonly rootCid: CID) {}

  /**
   * retrieves the merkle tree's metadata. The metadata must include the number of leaves in the tree
   * @returns promise for the metadata
   */
  async getMetadata(): Promise<Metadata> {
    if (!this._metadata) {
      this._metadata = await this.ipfsService.retrieveFromIPFS(this.rootCid, METADATA_PATH)
    }

    return this._metadata
  }

  /**
   * retreives the path to the leaf and the cid found at the provided index
   * @param index index of the leaf
   * @returns promise for the cid stored in the leaf
   */
  async getLeafData(index: number): Promise<{ cid: CID; path: PathDirection[] }> {
    const metadata = await this.getMetadata()
    const path = getPath(index, metadata.numEntries)

    const parent = await this.ipfsService.retrieveFromIPFS(
      this.rootCid,
      path.slice(0, -1).join('/')
    )
    const lastPathDirection = path[path.length - 1]
    const cid = parent[lastPathDirection].toString()

    return { cid, path }
  }
}
