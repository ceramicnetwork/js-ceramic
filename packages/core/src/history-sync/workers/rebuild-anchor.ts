import { default as PgBoss } from 'pg-boss'
import type { SendOptions } from 'pg-boss'
import type { AnchorProof, AnchorCommit, DiagnosticsLogger } from '@ceramicnetwork/common'
import { MerkleTreeLoader } from '../utils.js'
import { StreamID } from '@ceramicnetwork/streamid'
import {
  type IpfsService,
  type HandleCommit,
  type RebuildAnchorJobData,
  REBUILD_ANCHOR_JOB,
} from '../interfaces.js'
import type { Worker, Job } from '@ceramicnetwork/job-queue'
import { CID } from 'multiformats/cid'
import { pathString } from '@ceramicnetwork/anchor-utils'
import PQueue from 'p-queue'

// Up to 1024 streams could be present in an anchor
const IPFS_LOAD_CONCURRENCY = 16

const REBUILD_ANCHOR_JOB_OPTIONS: SendOptions = {
  retryLimit: 5,
  retryDelay: 60, // 1 minute
  retryBackoff: true,
  expireInHours: 12,
  retentionDays: 3,
}

export function createRebuildAnchorJob(
  proof: AnchorProof,
  models: string[],
  options: SendOptions = REBUILD_ANCHOR_JOB_OPTIONS
): Job<RebuildAnchorJobData> {
  return {
    name: REBUILD_ANCHOR_JOB,
    data: {
      models,
      chainId: proof.chainId,
      txHash: proof.txHash.toString(),
      root: proof.root.toString(),
      txType: proof.txType,
    },
    options,
  }
}

/**
 * Worker that recreates the anchor commits based on the anchor proof given.
 * It ensures that the data is stored and handled.
 */
export class RebuildAnchorWorker implements Worker<RebuildAnchorJobData> {
  constructor(
    private readonly ipfsService: IpfsService,
    private readonly handleCommit: HandleCommit,
    private readonly logger: DiagnosticsLogger
  ) {}

  private async getModelForStream(streamId: StreamID): Promise<StreamID | null> {
    const signedCommit = await this.ipfsService.retrieveCommit(
      streamId.cid, // genesis commit CID
      streamId
    )

    const genesisCommit = signedCommit?.link
      ? await this.ipfsService.retrieveCommit(signedCommit.link, streamId)
      : signedCommit

    if (!genesisCommit?.header?.model) {
      return null
    }

    return StreamID.fromBytes(genesisCommit.header.model)
  }

  /**
   * Takes an anchor proof, and stores it if it has no been stored.
   * Using the merkle root cid contained in the proof, it finds the leaf commits
   * that need to be synced, recreates their anchor commits, and handles them.
   * @param job job that contains an anchor proof in its data
   * @returns
   */
  async handler(job: PgBoss.Job) {
    const jobData = job.data as RebuildAnchorJobData
    const proof: AnchorProof = {
      chainId: jobData.chainId,
      txHash: CID.parse(jobData.txHash),
      root: CID.parse(jobData.root),
      txType: jobData.txType,
    }

    const proofCid = await this.ipfsService.storeRecord(proof as any).catch((err) => {
      this.logger.err(
        `Failed to store a proof on ipfs for root ${jobData.root} and txHash ${jobData.txHash} for models ${jobData.models} with error: ${err} `
      )
      // TODO (CDB-2291): add failure job for root cid
    })
    if (!proofCid) {
      return
    }

    const merkleTreeLeafLoader = new MerkleTreeLoader(this.ipfsService, proof.root)
    const metadata = await merkleTreeLeafLoader.getMetadata().catch((err) => {
      this.logger.err(
        `Failed to retrieve the merkle tree metadata for root ${jobData.root} and txHash ${jobData.txHash} for models ${jobData.models} with error: ${err} `
      )
      // TODO (CDB-2291): add failure job for root cid
    })
    if (!metadata) {
      return
    }

    const tasks = metadata.streamIds.map((stream, i) => {
      return async () => {
        try {
          const streamId = StreamID.fromString(stream)
          const model = await this.getModelForStream(streamId)

          const shouldIndex = model
            ? jobData.models.some((modelNeedingSync) => modelNeedingSync === model.toString())
            : false

          if (shouldIndex) {
            const { cid, path } = await merkleTreeLeafLoader.getLeafData(i)

            const anchorCommit: AnchorCommit = {
              id: streamId.cid,
              prev: cid,
              proof: proofCid,
              path: pathString(path),
            }

            const anchorCommitCid = await this.ipfsService.storeCommit(anchorCommit)

            await this.handleCommit(streamId, anchorCommitCid, model)

            this.logger.debug(
              `Successfully handled anchor commit ${anchorCommitCid} for stream ${streamId.toString()} and model ${model.toString()} using merkle tree root ${
                jobData.root
              }`
            )
          }
        } catch (err) {
          this.logger.err(
            `Failed to recreate the anchor commit for stream ${stream} using root ${jobData.root} and txHash ${jobData.txHash} for models ${jobData.models} with error: ${err} `
          )
          // TODO (CDB-2291): add failure job for streamId
        }
      }
    })

    const queue = new PQueue({ concurrency: IPFS_LOAD_CONCURRENCY })
    await queue.addAll(tasks)

    this.logger.debug(
      `Rebuild anchor job completed for models ${jobData.models}, root ${jobData.root}, and txHash ${jobData.txHash}`
    )
  }
}
