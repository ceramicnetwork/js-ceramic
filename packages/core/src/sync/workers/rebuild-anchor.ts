import { default as PgBoss } from 'pg-boss'
import type { AnchorProof, AnchorCommit } from '@ceramicnetwork/common'
import { MerkleTreeLoader } from '../utils.js'
import { StreamID } from '@ceramicnetwork/streamid'
import type { IpfsService, HandleCommit } from '../interfaces.js'
import type { Worker } from '../../state-management/job-queue.js'

interface RebuildAnchorJobData {
  proof: AnchorProof
  models: string[]
}

/**
 * Worker that recreates the anchor commits based on the anchor proof given.
 * It ensures that the data is stored and handled.
 */
export class RebuildAnchorWorker implements Worker {
  constructor(
    private readonly ipfsService: IpfsService,
    private readonly handleCommit: HandleCommit
  ) {}

  private async getModelForStream(streamId: StreamID): Promise<StreamID | null> {
    const signedCommit = await this.ipfsService.retrieveCommit(
      streamId.cid, // genesis commit CID
      streamId
    )

    const genesisCommit = await this.ipfsService.retrieveCommit(signedCommit.link, streamId)

    if (!genesisCommit.header.model) {
      return
    }

    return StreamID.fromBytes(genesisCommit.header.model)
  }

  /**
   * Takes an anchor proof, and stores it if it has no been stored.
   * Using the merkle root cid contained in the proof, it finds the leaf commits
   * that need to be synced, recreates their anchor commits, and handles them.
   * @param job job that contains an anchor proof in its data
   * @returns
   */
  async handler(job: PgBoss.Job) {
    const jobData = job.data as RebuildAnchorJobData
    const { proof, models } = jobData

    const proofCid = await this.ipfsService.storeRecord(proof as any).catch(() => {
      // TODO: add failure job for root cid
    })
    if (!proofCid) {
      return
    }

    const merkleTreeLeafLoader = new MerkleTreeLoader(this.ipfsService, proof.root)
    const metadata = await merkleTreeLeafLoader.getMetadata().catch(() => {
      // TODO: add failure job for root cid
    })
    if (!metadata) {
      return
    }

    const streams = metadata.streamIds

    for (let i = 0; i < streams.length; i++) {
      try {
        const streamId = StreamID.fromString(streams[i])

        const model = await this.getModelForStream(streamId)

        const shouldIndex = model
          ? models.some((modelNeedingSync) => modelNeedingSync === model.toString())
          : false

        if (shouldIndex) {
          const { cid, path } = await merkleTreeLeafLoader.getLeafData(i)

          const anchorCommit: AnchorCommit = {
            id: streamId.cid,
            prev: cid,
            proof: proofCid,
            path: path.join('/'),
          }

          const anchorCommitCid = await this.ipfsService.storeCommit(anchorCommit)

          await this.handleCommit(streamId, anchorCommitCid, model)
        }
      } catch (err) {
        // TODO: add failure job for streamId
      }
    }
  }
}
