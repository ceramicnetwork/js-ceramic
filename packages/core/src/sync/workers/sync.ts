import { default as PgBoss } from 'pg-boss'
import { type SendOptions } from 'pg-boss'
import { IJobQueue, Job, Worker } from '../../state-management/job-queue.js'
import { BlockProofs, createBlocksProofsLoader } from '@ceramicnetwork/anchor-listener'
import type { Provider } from '@ethersproject/providers'
import { concatMap, lastValueFrom, of, catchError } from 'rxjs'
import { createRebuildAnchorJob } from './rebuild-anchor.js'
import {
  RebuildAnchorJobData,
  JobData,
  HISTORY_SYNC_JOB,
  CONTINUOUS_SYNC_JOB,
} from '../interfaces.js'
import { DiagnosticsLogger } from '@ceramicnetwork/common'
import { SyncJobData } from '../interfaces.js'

const SYNC_JOB_OPTIONS: SendOptions = {
  retryLimit: 5,
  retryDelay: 60, // 1 minute
  retryBackoff: true,
  expireInHours: 12,
  retentionDays: 3,
}

export function createContinuousSyncJob(
  data: SyncJobData,
  options?: SendOptions
): Job<SyncJobData> {
  return {
    name: CONTINUOUS_SYNC_JOB,
    data,
    options: options || SYNC_JOB_OPTIONS,
  }
}

export function createHistorySyncJob(data: SyncJobData, options?: SendOptions): Job<SyncJobData> {
  return {
    name: HISTORY_SYNC_JOB,
    data,
    options: options || SYNC_JOB_OPTIONS,
  }
}

export interface SyncCompleteData {
  isHistoricSync: boolean
  modelId: string
}

/**
 * Worker that creates recreateAnchor jobs based on the anchor proofs  recreates the anchor commits based on the anchor proof given.
 * It ensures that the data is stored and handled.
 */
export class SyncWorker implements Worker<SyncJobData> {
  constructor(
    private readonly provider: Provider,
    private readonly jobQueue: IJobQueue<JobData>,
    private readonly chainId,
    private readonly logger: DiagnosticsLogger,
    private readonly syncCompleteCallback: (SyncCompleteData) => void
  ) {}

  /**
   * takes a sync job and retrieves the block proofs for each block between fromBlock - toBlock.
   * For each block with block proofs, create RebuildAnchorJobs which will recreate the anchor commits
   * and handle them.
   * @param job job that contains the sync job data
   */
  async handler(job: PgBoss.Job) {
    const jobData = job.data as SyncJobData
    const { fromBlock, toBlock, models } = jobData
    const currentBlock = jobData.currentBlock || fromBlock

    const blockProof$ = createBlocksProofsLoader({
      provider: this.provider,
      chainId: this.chainId,
      fromBlock: currentBlock,
      toBlock,
    }).pipe(
      // catch any errors so it doesn't stop any block proofs currently processing
      catchError((err) => {
        this.logger.err(
          `Received error when retrieving block proofs for models ${models} from block ${currentBlock} to block ${toBlock}: ${err}`
        )
        return of(null)
      }),
      // created rebuild anchor jobs for each block's proofs. Waits for last block to finish processing
      // before starting the next block
      concatMap(async (blockProofs: BlockProofs | null) => {
        if (!blockProofs) {
          throw Error('Error loading block proof')
        }

        const { proofs, blockNumber } = blockProofs
        if (proofs.length > 0) {
          const jobs: Job<RebuildAnchorJobData>[] = proofs.map((proof) =>
            createRebuildAnchorJob(proof, models)
          )

          await this.jobQueue.addJobs(jobs)

          this.logger.debug(
            `Successfully created ${jobs.length} rebuild anchor commit jobs for block ${blockNumber}`
          )
        }

        await this.jobQueue.updateJob(job.id, {
          fromBlock,
          currentBlock: blockNumber + 1,
          toBlock,
          models,
        })
      })
    )

    await lastValueFrom(blockProof$).then(() => {
      if (job.name == HISTORY_SYNC_JOB) {
        for (const model of models) {
          this.syncCompleteCallback({
            isHistoricSync: true,
            modelId: model,
          })
        }
      }
      this.logger.debug(
        `Sync completed for models ${models} from block ${fromBlock} to block ${toBlock}`
      )
    })
  }
}
