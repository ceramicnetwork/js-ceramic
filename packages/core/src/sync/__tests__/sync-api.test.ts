import { jest } from '@jest/globals'
import pgSetup from '@databases/pg-test/jest/globalSetup'
import pgTeardown from '@databases/pg-test/jest/globalTeardown'
import knex, { type Knex } from 'knex'
import { Observable } from 'rxjs'
import { LoggerProvider } from '@ceramicnetwork/common'

import { REBUILD_ANCHOR_JOB_NAME, SYNC_JOB_NAME } from '../interfaces.js'
import { RebuildAnchorWorker } from '../workers/rebuild-anchor.js'
import { SyncWorker, createSyncJob } from '../workers/sync.js'

const createBlockProofsListener = jest.fn(() => new Observable())

jest.unstable_mockModule('@ceramicnetwork/anchor-listener', () => {
  return { createBlockProofsListener }
})

describe('Sync API', () => {
  jest.setTimeout(150000) // 2.5mins timeout for initial docker fetch+init
  let dbConnection: Knex

  async function dropTables() {
    const { STATE_TABLE_NAME } = await import('../sync-api.js')
    await dbConnection.schema.dropTableIfExists(STATE_TABLE_NAME)
  }

  const logger = new LoggerProvider().getDiagnosticsLogger()

  beforeAll(async () => {
    await pgSetup()
  })

  beforeEach(() => {
    dbConnection = knex({
      client: 'pg',
      connection: process.env.DATABASE_URL,
    })
  })

  afterEach(async () => {
    await dropTables()
    await dbConnection.destroy()
  })

  afterAll(async () => {
    await pgTeardown()
  })

  test('_initJobQueue() initializes the JobQueue instance with the workers', async () => {
    const { SyncApi } = await import('../sync-api.js')
    const sync = new SyncApi(
      { chainId: 'eip155:1337', db: process.env.DATABASE_URL as string },
      {} as any,
      {} as any,
      {} as any,
      {} as any,
      {} as any
    )

    const init = jest.fn()
    // @ts-ignore private field
    sync.jobQueue = { init }

    await sync._initJobQueue()
    expect(init).toHaveBeenCalledWith({
      [REBUILD_ANCHOR_JOB_NAME]: expect.any(RebuildAnchorWorker),
      [SYNC_JOB_NAME]: expect.any(SyncWorker),
    })
  })

  test('_initModelsToSync() loads the models to sync from DB and adds them to the local set', async () => {
    const expectedModels = ['abc123', 'abc456', 'abc789', 'def123']
    const indexedModels = jest.fn(() => Promise.resolve(expectedModels))

    const { SyncApi } = await import('../sync-api.js')
    const sync = new SyncApi(
      { chainId: 'eip155:1337', db: process.env.DATABASE_URL as string },
      {} as any,
      {} as any,
      {} as any,
      { indexedModels } as any,
      {} as any
    )

    await sync._initModelsToSync()
    expect(indexedModels).toHaveBeenCalled()
    expect(Array.from(sync.modelsToSync)).toEqual(expectedModels)
  })

  describe('_initStateTable()', () => {
    test('creates the table if not existing', async () => {
      const { STATE_TABLE_NAME, SyncApi } = await import('../sync-api.js')
      const sync = new SyncApi(
        { chainId: 'eip155:1337', db: process.env.DATABASE_URL as string },
        {} as any,
        {} as any,
        {} as any,
        {} as any,
        {} as any
      )
      await expect(sync._initStateTable()).resolves.toEqual({})
      await expect(dbConnection.from(STATE_TABLE_NAME).first()).resolves.toEqual({
        processed_block_hash: null,
        processed_block_number: null,
      })
    })

    test('reads the state from the table if existing', async () => {
      const { STATE_TABLE_NAME, SyncApi } = await import('../sync-api.js')
      await dbConnection.schema.createTable(STATE_TABLE_NAME, function (table) {
        table.string('processed_block_hash', 1024)
        table.integer('processed_block_number')
      })
      await dbConnection
        .into(STATE_TABLE_NAME)
        .insert({ processed_block_hash: '0x123abc', processed_block_number: 10 })

      const sync = new SyncApi(
        { chainId: 'eip155:1337', db: process.env.DATABASE_URL as string },
        {} as any,
        {} as any,
        {} as any,
        {} as any,
        {} as any
      )
      await expect(sync._initStateTable()).resolves.toEqual({
        processedBlockHash: '0x123abc',
        processedBlockNumber: 10,
      })
    })
  })

  test('_initBlockSubscription()', async () => {
    const { BLOCK_CONFIRMATIONS, SyncApi } = await import('../sync-api.js')
    const provider = {} as any

    const sync = new SyncApi(
      { chainId: 'eip155:1337', db: process.env.DATABASE_URL as string },
      {} as any,
      {} as any,
      provider,
      {} as any,
      {} as any
    )
    sync._initBlockSubscription('abc123')
    expect(createBlockProofsListener).toHaveBeenCalledWith({
      confirmations: BLOCK_CONFIRMATIONS,
      chainId: 'eip155:1337',
      provider: provider,
      expectedParentHash: 'abc123',
    })
    // @ts-ignore private field
    expect(sync.subscription).toBeDefined()
  })

  describe('init() initializes sync', () => {
    test('calls the internal initialization methods and retrieves the first block to sync from', async () => {
      const { BLOCK_CONFIRMATIONS, SyncApi } = await import('../sync-api.js')
      const getBlock = jest.fn(() => ({ number: 10, hash: 'abc123' }))
      const sync = new SyncApi(
        { chainId: 'eip155:1337', db: process.env.DATABASE_URL as string },
        {} as any,
        {} as any,
        { getBlock } as any,
        {} as any,
        {} as any
      )

      const initStateTable = jest.fn(() => ({ processedBlockNumber: 10 }))
      const initModelsToSync = jest.fn()
      const initJobQueue = jest.fn()
      const initBlockSubscription = jest.fn()
      sync._initStateTable = initStateTable as any
      sync._initModelsToSync = initModelsToSync as any
      sync._initJobQueue = initJobQueue as any
      sync._initBlockSubscription = initBlockSubscription as any

      await sync.init()
      expect(getBlock).toHaveBeenCalledWith(-BLOCK_CONFIRMATIONS)
      expect(initStateTable).toHaveBeenCalled()
      expect(initModelsToSync).toHaveBeenCalled()
      expect(initJobQueue).toHaveBeenCalled()
      expect(initBlockSubscription).toHaveBeenCalledWith('abc123')

      await sync.shutdown()
    })

    test('adds a job to do a full sync if there is no previously processed block', async () => {
      const { INITIAL_INDEXING_BLOCK, SyncApi } = await import('../sync-api.js')

      const getBlock = jest.fn(() => ({ number: 10, hash: 'abc123' }))
      const expectedModels = ['abc123', 'abc456', 'abc789', 'def123']
      const indexedModels = jest.fn(() => Promise.resolve(expectedModels))

      const sync = new SyncApi(
        { chainId: 'eip155:1337', db: process.env.DATABASE_URL as string },
        {} as any,
        {} as any,
        { getBlock } as any,
        { indexedModels } as any,
        {} as any
      )

      const initStateTable = jest.fn(() => ({ processedBlockNumber: null }))
      sync._initStateTable = initStateTable as any
      const addSyncJob = jest.fn()
      sync._addSyncJob = addSyncJob as any

      await sync.init()
      expect(addSyncJob).toHaveBeenCalledWith({
        fromBlock: INITIAL_INDEXING_BLOCK,
        toBlock: 10,
        models: expectedModels,
      })

      await sync.shutdown()
    })

    test('adds a job to sync from the previously processed block', async () => {
      const { SyncApi } = await import('../sync-api.js')

      const getBlock = jest.fn(() => ({ number: 10, hash: 'abc123' }))
      const expectedModels = ['abc123', 'abc456', 'abc789', 'def123']
      const indexedModels = jest.fn(() => Promise.resolve(expectedModels))

      const sync = new SyncApi(
        { chainId: 'eip155:1337', db: process.env.DATABASE_URL as string },
        {} as any,
        {} as any,
        { getBlock } as any,
        { indexedModels } as any,
        {} as any
      )

      const initStateTable = jest.fn(() => ({ processedBlockNumber: 5 }))
      sync._initStateTable = initStateTable as any
      const addSyncJob = jest.fn()
      sync._addSyncJob = addSyncJob as any

      await sync.init()
      expect(addSyncJob).toHaveBeenCalledWith({
        fromBlock: 5,
        toBlock: 10,
        models: expectedModels,
      })

      await sync.shutdown()
    })

    test('does not add a job if already in sync', async () => {
      const { SyncApi } = await import('../sync-api.js')
      const getBlock = jest.fn(() => ({ number: 10, hash: 'abc123' }))
      const expectedModels = ['abc123', 'abc456', 'abc789', 'def123']
      const indexedModels = jest.fn(() => Promise.resolve(expectedModels))

      const sync = new SyncApi(
        { chainId: 'eip155:1337', db: process.env.DATABASE_URL as string },
        {} as any,
        {} as any,
        { getBlock } as any,
        { indexedModels } as any,
        {} as any
      )

      const initStateTable = jest.fn(() => ({ processedBlockNumber: 10 }))
      sync._initStateTable = initStateTable as any
      const addSyncJob = jest.fn()
      sync._addSyncJob = addSyncJob as any

      await sync.init()
      expect(addSyncJob).not.toHaveBeenCalled()

      await sync.shutdown()
    })
  })

  test('shutdown() stops the anchor subscription and job queue', async () => {
    const { SyncApi } = await import('../sync-api.js')
    const sync = new SyncApi(
      { chainId: 'eip155:1337', db: process.env.DATABASE_URL as string },
      {} as any,
      {} as any,
      {} as any,
      {} as any,
      {} as any
    )

    const unsubscribe = jest.fn()
    // @ts-ignore private field
    sync.subscription = { unsubscribe }
    const stop = jest.fn()
    // @ts-ignore private field
    sync.jobQueue = { stop }

    await sync.shutdown()
    expect(unsubscribe).toHaveBeenCalled()
    expect(stop).toHaveBeenCalled()
  })

  describe('addModelSync() adds a model or models to sync', () => {
    test('handles a single model as input', async () => {
      const { SyncApi } = await import('../sync-api.js')
      const sync = new SyncApi(
        { chainId: 'eip155:1337', db: process.env.DATABASE_URL as string },
        {} as any,
        {} as any,
        {} as any,
        {} as any,
        {} as any
      )

      const addSyncJob = jest.fn()
      sync._addSyncJob = addSyncJob as any

      const data = { fromBlock: 1, toBlock: 10, models: ['abc123'] }
      await sync.startModelSync(data.fromBlock, data.toBlock, 'abc123')
      expect(addSyncJob).toHaveBeenCalledWith(data)
      expect(Array.from(sync.modelsToSync)).toEqual(data.models)
    })

    test('handles multiple models as input', async () => {
      const { SyncApi } = await import('../sync-api.js')
      const sync = new SyncApi(
        { chainId: 'eip155:1337', db: process.env.DATABASE_URL as string },
        {} as any,
        {} as any,
        {} as any,
        {} as any,
        {} as any
      )

      const addSyncJob = jest.fn()
      sync._addSyncJob = addSyncJob as any

      const data = { fromBlock: 1, toBlock: 10, models: ['abc123', 'def456'] }
      await sync.startModelSync(data.fromBlock, data.toBlock, data.models)
      expect(addSyncJob).toHaveBeenCalledWith(data)
      expect(Array.from(sync.modelsToSync)).toEqual(data.models)
    })
  })

  test('_addSyncJob() creates a sync job and adds it to the queue', async () => {
    const { SyncApi } = await import('../sync-api.js')
    const sync = new SyncApi(
      { chainId: 'eip155:1337', db: process.env.DATABASE_URL as string },
      {} as any,
      {} as any,
      {} as any,
      {} as any,
      {} as any
    )

    const addJob = jest.fn()
    // @ts-ignore private field
    sync.jobQueue = { addJob }

    const data = { fromBlock: 1, toBlock: 10, models: ['abc123', 'abc456'] }
    await sync._addSyncJob(data)
    expect(addJob).toHaveBeenCalledWith(createSyncJob(data))
  })

  test('_updateStoredState() updates the state in DB', async () => {
    const { STATE_TABLE_NAME, SyncApi } = await import('../sync-api.js')
    const sync = new SyncApi(
      { chainId: 'eip155:1337', db: process.env.DATABASE_URL as string },
      {} as any,
      {} as any,
      {} as any,
      {} as any,
      {} as any
    )
    await sync._initStateTable()
    // Check state before update
    await expect(dbConnection.from(STATE_TABLE_NAME).first()).resolves.toEqual({
      processed_block_hash: null,
      processed_block_number: null,
    })
    await sync._updateStoredState({
      processedBlockHash: '0x123abc',
      processedBlockNumber: 10,
    })
    await expect(dbConnection.from(STATE_TABLE_NAME).first()).resolves.toEqual({
      processed_block_hash: '0x123abc',
      processed_block_number: 10,
    })
  })
})
