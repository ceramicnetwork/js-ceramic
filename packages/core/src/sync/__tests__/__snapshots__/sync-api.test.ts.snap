// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sync API _createJobStatus 1`] = `
"Logging state of running ComposeDB syncs
 {
   \\"activeSyncs\\": [
      {
         \\"currentBlock\\": 101,
         \\"startBlock\\": 100,
         \\"endBlock\\": 200,
         \\"createdAt\\": \\"2023-02-21T21:43:00.491Z\\",
         \\"startedAt\\": \\"2023-02-21T21:44:40.491Z\\"
      }
   ],
   \\"continuousSync\\": [
      {
         \\"confirmations\\": 20,
         \\"currentBlock\\": 500
      }
   ],
   \\"pendingSyncs\\": [
      {
         \\"startBlock\\": 100,
         \\"endBlock\\": 200,
         \\"createdAt\\": \\"2023-02-21T21:43:00.491Z\\"
      }
   ]
}"
`;
